test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,442 @@
+
+        // Wasm only supports conversions to integer with truncation rounding direction:
+        // testfloat_gen needs -rminMag option.
+        // Conversion of a floating-point number to an integer format, when the source is NaN,
+        // infinity, or a value that would convert to an integer outside the range of the result
+        // format under the applicable rounding attribute.
+        {"f32_trunc_to_i32", {24, Type::i32, Type::f32, Options::TrapIsInvalidOperation}},
+        {"f32_trunc_to_ui32", {25, Type::i32, Type::f32, Options::TrapIsInvalidOperation}},
+        {"f32_trunc_to_i64", {26, Type::i64, Type::f32, Options::TrapIsInvalidOperation}},
+        {"f32_trunc_to_ui64", {27, Type::i64, Type::f32, Options::TrapIsInvalidOperation}},
+        {"f64_trunc_to_i32", {28, Type::i32, Type::f64, Options::TrapIsInvalidOperation}},
+        {"f64_trunc_to_ui32", {29, Type::i32, Type::f64, Options::TrapIsInvalidOperation}},
+        {"f64_trunc_to_i64", {30, Type::i64, Type::f64, Options::TrapIsInvalidOperation}},
+        {"f64_trunc_to_ui64", {31, Type::i64, Type::f64, Options::TrapIsInvalidOperation}},
+
+        {"f32_to_f64", {32, Type::f64, Type::f32}},
+        {"f64_to_f32", {33, Type::f32, Type::f64}},
+
+        {"f32_ceil", {34, Type::f32, Type::f32}},
+        {"f32_floor", {35, Type::f32, Type::f32}},
+        {"f32_trunc", {36, Type::f32, Type::f32}},
+        {"f32_nearest", {37, Type::f32, Type::f32}},
+        {"f64_ceil", {38, Type::f64, Type::f64}},
+        {"f64_floor", {39, Type::f64, Type::f64}},
+        {"f64_trunc", {40, Type::f64, Type::f64}},
+        {"f64_nearest", {41, Type::f64, Type::f64}},
+    };
+
+    if (const auto it = map.find(name); it != map.end())
+        return it->second;
+    throw std::invalid_argument{"unknown <function>: " + std::string{name}};
+}
+
+constexpr fizzy::Value make_value(Type type, uint64_t bits) noexcept
+{
+    switch (type)
+    {
+    case Type::i32:
+        return fizzy::Value{static_cast<uint32_t>(bits)};
+    case Type::i64:
+        return fizzy::Value{bits};
+    case Type::f32:
+        return fizzy::Value{FP{static_cast<uint32_t>(bits)}.value};
+    case Type::f64:
+        return fizzy::Value{FP{bits}.value};
+    }
+    __builtin_unreachable();
+}
+
+class TypedValue
+{
+    Type m_type;
+    fizzy::Value m_value;
+
+public:
+    constexpr TypedValue(Type type, fizzy::Value value) noexcept : m_type{type}, m_value{value} {}
+
+    constexpr TypedValue(Type type, uint64_t bits) noexcept
+      : m_type{type}, m_value{make_value(type, bits)}
+    {}
+
+    bool eq(uint64_t expected_bits, bool ignore_nans) const
+    {
+        switch (m_type)
+        {
+        case Type::i32:
+        case Type::i64:
+            return m_value.i64 == expected_bits;
+        case Type::f32:
+        {
+            const FP value{m_value.f32};
+            const FP expected{static_cast<uint32_t>(expected_bits)};
+            if (ignore_nans && std::isnan(expected.value))
```

Comment:
So what does `ignore_nans` means? To ignore payloads?

Reply:
Yes.